tests/e2e/utils/versions_test.go

```diff
@@ -0,0 +1,348 @@
+// Copyright (C) 2022, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package utils
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"strconv"
+	"testing"
+
+	"github.com/ava-labs/avalanche-cli/pkg/application"
+	"github.com/ava-labs/avalanche-cli/pkg/models"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/stretchr/testify/assert"
+	"golang.org/x/mod/semver"
+)
+
+type testContext struct {
+	expected    map[string]string
+	sourceEVM   string
+	spacesVM    string
+	sourceAvago string
+	shouldFail  bool
+}
+
+type testMapper struct {
+	app            *application.Avalanche
+	currentContext *testContext
+	srv            *httptest.Server
+	t              *testing.T
+}
+
+func newTestMapper(t *testing.T) *testMapper {
+	app := &application.Avalanche{
+		Downloader: application.NewDownloader(),
+		Log:        logging.NoLog{},
+	}
+	return &testMapper{
+		app,
+		nil,
+		nil,
+		t,
+	}
+}
+
+func (m *testMapper) GetLatestAvagoByProtoVersion(app *application.Avalanche, rpcVersion int, url string) (string, error) {
+	cBytes := []byte(m.currentContext.sourceAvago)
+
+	var compat models.AvagoCompatiblity
+	if err := json.Unmarshal(cBytes, &compat); err != nil {
+		return "", err
+	}
+	vers := compat[strconv.Itoa(rpcVersion)]
+	if len(vers) == 0 {
+		return "", errors.New("test zero length versions")
+	}
+	if len(vers) > 1 {
+		semver.Sort(vers)
+	}
+	// take the last
+	return vers[len(vers)-1], nil
+}
+
+func (m *testMapper) getVersionMapping(tc *testContext) (map[string]string, error) {
+	mapping = nil
+	m.currentContext = tc
+	return GetVersionMapping(m)
+}
+
+func (m *testMapper) GetApp() *application.Avalanche {
+	return m.app
+}
+
+func (m *testMapper) GetCompatURL(vmType models.VMType) string {
+	switch vmType {
+	case models.SubnetEvm:
+		return m.srv.URL + "/evm"
+	case models.SpacesVM:
+		return m.srv.URL + "/spaces"
+	default:
+		m.t.Fatalf("unexpected vmType: %T", vmType)
+	}
+	return ""
+}
+
+func (m *testMapper) GetAvagoURL() string {
+	return m.srv.URL + "/avago"
+}
+
+func (m *testMapper) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	w.Header().Add("Content-Type", "application/json")
+	var err error
+	switch r.URL.Path {
+	case "/evm":
+		_, err = w.Write([]byte(m.currentContext.sourceEVM))
+	case "/spaces":
+		_, err = w.Write([]byte(m.currentContext.spacesVM))
+	case "/avago":
+		_, err = w.Write([]byte(m.currentContext.sourceAvago))
+	default:
+		m.t.Fatalf("Unexpected path URL for test server: %s\n", r.URL.Path)
+	}
+	if err != nil {
+		m.t.Fatal(err)
+	}
+}
+
+func TestGetVersionMapping(t *testing.T) {
```

Comment:
Sorry, this isn't going to be the most helpful review comment ever, but I am extremely confused as to what this file is testing. I would suggest labeling/naming each test case and moving some of the above logic into the test so that it looks a bit more linear. It's very hard to follow.

Reply:
I added detailed code documentation to the test.
The idea of the test is that it actually runs the `GetVersionMapping`. For that to be useful, it needs to download some data. The test fakes github by returning `compatibility.json` content from the test, and makes sure that the return from `GetVersionMapping` matches an expected set.